CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+
+        private ICollection<(CacheItemRemovedEventArgs Args, DateTime Time)> _onRemoveByHandleInvocations;
+
+        [OneTimeSetUp]
+        public void Setup()
+        {
+            _serverMock = new Mock<IServer>();
+            _databaseMock = new Mock<IDatabase>();
+            _connectionMock = new Mock<IConnectionMultiplexer>();
+
+            _fauxDatabase = new ConcurrentDictionary<RedisKey, (RedisValue[], DateTime)>();
+
+            SetupServerMock();
+            SetupDatabaseMock();
+            SetupConnectionMock();
+            SetupCache();
+        }
+
+        #region Setup Methods
+
+        private void SetupServerMock()
+        {
+            _serverMock.SetupGet(server => server.IsConnected).Returns(true);
+            _serverMock.SetupGet(server => server.Features).Returns(new RedisFeatures(Version.Parse("3.0.504")));
+            _serverMock
+                .Setup(server => server.Keys(It.IsAny<int>(), It.IsAny<RedisValue>(), It.IsAny<int>(), It.IsAny<long>(),
+                    It.IsAny<int>(), It.IsAny<CommandFlags>()))
+                .Returns<int, RedisValue, int, long, int, CommandFlags>(
+                    (db, pattern, pageSize, cursor, pageOffset, flags) => _fauxDatabase.Keys);
+        }
+
+        private void SetupDatabaseMock()
+        {
+            _databaseMock.SetupGet(database => database.Database).Returns(0);
+            _databaseMock.Setup(database => database.KeyIdleTime(It.IsAny<RedisKey>(), It.IsAny<CommandFlags>()))
+                .Returns<RedisKey, CommandFlags>((key, flags) => DateTime.Now - _fauxDatabase[key].InsertionTime);
+
+            void Put(byte[] hash, RedisKey[] keys, RedisValue[] values, CommandFlags flags)
+            {
+                RedisValue[] newValues = {values[0], values[2], values[3], values[4], values[1], values[5], true};
+                (RedisValue[], DateTime) tuple = (newValues, DateTime.Now);
+
+                RedisKey key = keys[0];
+
+                if (_fauxDatabase.ContainsKey(key))
+                {
+                    _fauxDatabase[key] = tuple;
+                }
+                else
+                {
+                    _fauxDatabase.Add(key, tuple);
+                }
+            }
+
+            // Put
+            _databaseMock
+                .Setup(database => database.ScriptEvaluate(It.IsAny<byte[]>(), It.IsAny<RedisKey[]>(),
+                    It.Is<RedisValue[]>(values => values != null), It.IsAny<CommandFlags>()))
+                .Callback<byte[], RedisKey[], RedisValue[], CommandFlags>(Put);
+
+            // Get
+            _databaseMock
+                .Setup(database => database.ScriptEvaluate(It.IsAny<byte[]>(), It.IsAny<RedisKey[]>(),
+                    It.Is<RedisValue[]>(values => values == null), It.IsAny<CommandFlags>()))
+                .Returns<byte[], RedisKey[], RedisValue[], CommandFlags>((hash, keys, values, flags) =>
+                    RedisResult.Create(_fauxDatabase[keys[0]].Values));
```

Comment:
Explain the `ScriptEvaluate` setup here. You can combine this and the previous explanation togeher. Like: "RedisCacheManager depends on StackExchange, which use ScriptEvaluate for acting against the Redis. As part of it, we:
1. need to make sure it got all of its parameters in order to work, so we pass the right values in the right order
2. setup callback and result of running scripts, in order to mock the behavior of cache in our fauxDatabase"
😄

Reply:
Added, tell me if you think more I should elaborate more